File: examples/Java/src/main/java/com/imatix/zguide/ex1_8/TaskVent.java
package com.imatix.zguide.ex1_8;

import org.zeromq.ZMQ.Context;
import org.zeromq.ZMQ.Socket;

import java.io.IOException;
import java.util.Random;

import static com.imatix.zguide.commons.ZConstants.URL.TCP_5557;
import static com.imatix.zguide.commons.ZConstants.URL.TCP_5558;
import static java.lang.String.format;
import static org.zeromq.ZMQ.PUSH;
import static org.zeromq.ZMQ.context;

/**
 * Task ventilator in Java
 * Binds PUSH socket to tcp://localhost:5557
 * Sends batch of tasks to workers via that socket
 *
 * @since 1.0
 */

public class TaskVent {

    public static void main(String[] args) throws IOException, InterruptedException {
        Context context = context(1);

        //  Socket to send messages on
        Socket sender = context.socket(PUSH);
        sender.bind(TCP_5557);
        //  Socket to send messages on
        Socket sink = context.socket(PUSH);
        sink.connect(TCP_5558); // Why connect ?
        System.out.println("Press Enter as soon as the workers are all set: ");
        System.in.read();
        System.out.println("Sending Tasks to the workers ...");
        //  The first message is "0" and signals start of batch
        sink.send("0", 0);
        //  Initialize random number generator
        Random random = new Random(System.currentTimeMillis());
        //  Send 100 tasks
        int totalDuration = 0;
        for (int taskNum = 0; taskNum < 100; taskNum++) {
            int workload;
            //  Random workload from 1 to 100 msecs.
            workload = random.nextInt(100) + 1;
            totalDuration += workload;
            System.out.println(workload + ".");
            String taskDuration = format("%d", workload);
            sender.send(taskDuration, 0);
        }

        System.out.printf("\nTotal expected cost: %d msec.", totalDuration);
        //  Give 0MQ time to deliver
        Thread.sleep(1000);
        sink.close();
        sender.close();
        context.term();
    }
}

File: examples/Free Pascal/taskvent.pas
{
  Task ventilator
  Binds PUSH socket to tcp://localhost:5557
  Sends batch of tasks to workers via that socket
  @author cpicanco <cpicanco@ufpa.br>
}
program taskvent;

{$MODE objfpc}{$H+}{$COPERATORS ON}

uses SysUtils, zmq, zmq.helpers;

var
  context, sender, sink : Pointer;
  task_nbr : integer;
  total_msec : integer = 0; //  Total expected cost in msecs
  workload : integer;

  LString : string;
begin
  context := zmq_ctx_new;

  //  Socket to send messages on
  sender := zmq_socket(context, ZMQ_PUSH);
  zmq_bind(sender, 'tcp://*:5557');

  //  Socket to send start of batch message on
  sink := zmq_socket(context, ZMQ_PUSH);
  zmq_connect(sink, 'tcp://localhost:5558');

  WriteLn('Press Enter when the workers are ready: ');
  ReadLn;
  WriteLn('Sending tasks to workers…');

  //  The first message is "0" and signals start of batch
  s_send(sink, '0');

  //  Initialize random number generator
  Randomize;

  //  Send 100 tasks
  for task_nbr := 0 to 100 -1 do
    begin
      //  Random workload from 1 to 100msecs
      workload := randof(100) + 1;
      total_msec += workload;
      LString := IntToStr(workload);
      s_send(sender, LString);
    end;

  WriteLn(Format('Total expected cost: %d msec', [total_msec]));

  zmq_close(sink);
  zmq_close(sender);
  zmq_ctx_destroy(context);
end.

